src/generators/ast.ts

```diff
@@ -198,10 +204,97 @@ function copyAdvancedKeywords(from: JSONSchema7, to: Schema): void {
   }
 }
 
+type ResolvedTypes = {
+  [key: string]: Schema;
+};
+
+let definedTypes: ResolvedTypes = {};
+
+export interface CustomType {
+  name: string;
+  schema: Schema;
+}
+
+export const customTypesByEvent: Record<string, CustomType[]> = {};
+
+export function extractCustomTypes(schema: JSONSchema7, eventName?: string): void {
```

Comment:
This is not a good function definition, you are setting `customTypesByEvent` which is exported in the module. The data returned should be contained within the function.

I would also set the return type Record<string, CustomType[]> as mandatory on the function and return it empty in case nothing can be extracted from the schema.

In this function, eventName will always be provided right ?
Based on the usage in the PR, it seems like eventName is always set

Reply:
Let's discuss more on this function when we sync up ?

Reply:
refactored